keylime/tpm/tpm_util.py

```diff
@@ -0,0 +1,156 @@
+import struct
+from typing import Dict, List, Tuple, Union
+
+from cryptography.exceptions import InvalidSignature
+from cryptography.hazmat import backends
+from cryptography.hazmat.primitives import hashes, serialization
+from cryptography.hazmat.primitives.asymmetric import ec, padding
+from cryptography.hazmat.primitives.asymmetric.ec import EllipticCurvePublicKey
+from cryptography.hazmat.primitives.asymmetric.rsa import RSAPublicKey
+from cryptography.hazmat.primitives.asymmetric.utils import Prehashed
+
+from keylime import keylime_logging
+from keylime.tpm import tpm2_objects
+
+logger = keylime_logging.init_logging("tpm_util")
+
+SupportedKeyTypes = Union[RSAPublicKey, EllipticCurvePublicKey]
+
+
+def verify(pubkey: SupportedKeyTypes, sig: bytes, digest: bytes, hashfunc: hashes.HashAlgorithm) -> None:
+    """Do signature verification with the given public key"""
+    if isinstance(pubkey, RSAPublicKey):
+        pubkey.verify(sig, digest, padding.PKCS1v15(), Prehashed(hashfunc))
+    elif isinstance(pubkey, EllipticCurvePublicKey):
+        pubkey.verify(sig, digest, ec.ECDSA(Prehashed(hashfunc)))
+
+
+def __get_pcrs_from_blob(pcrblob: bytes) -> Tuple[int, Dict[int, int], List[bytes]]:
+    """This function is specific to the Intel tools with data in little endian format.
+    Data structures were not marshalled but written as whole out of memory."""
```

Comment:
This can be a problem. We currently have the Rust agent in Intel, S390X and AArch64, that shares different endianness (little, big, bi)

I found issues in the tpm2-tools in the past (and present), so I am not sure if this was broken when using tpm2_quote, but this can comeback as a problem later.

Reply:
Yes, once we do major changes we should move to a more universal format instead of using the binary blob from tpm2-tools.

Reply:
First step is to replace the tpm2-tools code with python code while maintaining compatibility. 2nd step is to **maybe** extend the format but probably still maintain compatibility between agent and verifier.

Reply:
I second @stefanberger here. This is an incremental, yet important, step in the right direction

Reply:
> I second @stefanberger here. This is an incremental, yet important, step in the right direction

I half disagree.

This change is introducing a backward compatibility issue in the code. Big endian architectures gets now unsupported because of the new parser.

I am not super familiar with the Intel tss code, but when I worked on it I saw bugs in the [de]marshaling code (honest bugs, miss interpretations on the standard, retro compatibility, etc). That for me is an indication that this is not a trivial task.

@stefanberger code is remarkable, and moving out of tpm2_tools CLI can be a good step, but re-implementing the parsers is not. If possible, I would ask pytss to do the parsing, as it will delegate in into the tss stack.

Reply:
> This change is introducing a backward compatibility issue in the code. Big endian architectures gets now unsupported because of the new parser.

That is not true. 
The pcrblob was always written in little native format and per the changes I made back then they are now written in little endian format on all platforms. The function `__get_pcrs_from_blob()` explicitly parses the data in little endian format using the '<' specifier.
The aikblob is the PEM encoded public key. No issues with endianess here.
The sigblob is written in big endian and parsed as such.
The quoteblob is written in big endian and parsed as such.

> @stefanberger code is remarkable, and moving out of tpm2_tools CLI can be a good step, but re-implementing the parsers is not. If possible, I would ask pytss to do the parsing, as it will delegate in into the tss stack.

If we rely on packages available across all the distros we want to run this on, we will have to wait. If we can rely on pip I **may** have a look.

Reply:
I was referring to these PR here for tpm2-tools that did some of the changes converting from native to little endian format for at least the quote aspects (read and write) of the tools. Using little endian, where native indian was used before, was the least disruptive for interoperability with existing x86/ppc64/arm systems and resolved it for s390. The rust agent TPM 2 library for sure inherited that format (as we know due to interoperability with the tpm2 tools we are about to get rid of) and probably also took over explicit writing in little endian :
- https://github.com/tpm2-software/tpm2-tools/pull/2228
- https://github.com/tpm2-software/tpm2-tools/pull/2224
- also: https://github.com/tpm2-software/tpm2-tools/pull/2230

Reply:
> That is not true.
> The pcrblob was always written in little (native on x86) format and per the changes I made back then they are now written in little endian format on all platforms.

My apologies then.

AFAICT the tools are still not fully endian agnostics:

https://github.com/tpm2-software/tpm2-tools/issues/3060
https://github.com/tpm2-software/tpm2-tools/pull/3184

so my worries was in that direction, as some fails were related with the checkquote test code.

Reply:
Makes me think we shouldn't rely on it...

Reply:
IOW, with this PR we control more of the few TPM related pieces we need.

Reply:
so, we all agree that merging this code will not bring a backward compatibility issue?

Reply:
> so, we all agree that merging this code will not bring a backward compatibility issue?

I think we can agree on this!

I still think that parsing the binary blob in here is doing that in the wrong level of abstraction.  But if I am the only one thinking that, I will not oppose to the merge, of course.

Reply:
> I still think that parsing the binary blob in here is doing that in the wrong level of abstraction.

I did not see code in the pytss project that would parse that pcrblob (which we now also write in Rust).

Reply:
> I did not see code in the pytss project that would parse that pcrblob (which we now also write in Rust).

... that is also true, the format from tpm2_tools seems to be very adhoc ...

I consider myself defeated! : )